packages/server/src/websockets/builder.ts

```diff
@@ -100,11 +102,24 @@ export default class BuilderSocket extends BaseSocket {
     })
   }
 
-  emitTableUpdate(ctx: any, table: Table) {
-    this.emitToRoom(ctx, ctx.appId, BuilderSocketEvent.TableChange, {
-      id: table._id,
-      table,
-    })
+  emitTableUpdate(ctx: any, table: Table, options?: EmitOptions) {
+    // This was added to make sure that sourceId is always present when
+    // sending this message to clients. Without this, tables without a
+    // sourceId (e.g. ta_users) won't get correctly updated client-side.
+    if (isInternalTable(table._id!)) {
+      table = processInternalTable(table)
+    }
```

Comment:
Two important changes here:

1. The addition of `EmitOptions` to indicate when we want a websocket update to also be processed by the event originator (necessary to update the state of the user table in the UI).
2. The enriching of the `table` with `processInternalTable`. This adds `sourceId` to the table, without which the UI won't use it to update its state.

Reply:
Happy to leave this in here, but I think this maybe needs another look. With this enrichment, and a new app, the users table *will not* have a sourceId when fetched from the tables list endpoint (as it does not actually exist in the doc), but then *will* have one when a table change event is broadcast. That means that the original frontend issue of not being able to create a relationship to the users table will still exist, and will only be fixed if the client receives a table change event with this enriched property.

So I think the sourceId issue is a totally separate thing.

@mike12345567 could you confirm if removing the sourceId property from users table in new apps is intended? If so then I'm leaning towards removing this enrichment here, and instead updating the frontend to understand that it's optional.

Reply:
The `sourceId` was really only designed for external datasources - but internal tables had one thrown on just for consistency - if the frontend expects it to be there for all tables then we can update the typing to make it so that it is non-optional on all tables - right now it is required for external but is optional for internal tables, however we can make that required I think.

@samwho - let me know if you'd like a hand with that!

Reply:
I just tried creating a brand new app and inspecting the return values from some API endpoints and here's what I found:

1. `GET /api/tables` returns a `sourceId` (tacked on, the `sourceId` is not present in Couch) on each table
2. `GET /api/tables/ta_users` does not return a `sourceId`
3. `GET /api/tables/ta_<id>` _does_ return a `sourceId` for another internal table (checking Fauxton, this appears to be because the table document has a `sourceId`)

@mike12345567 I'd appreciate a hand with this, thank you 🙏